langlib/lang.query/src/main/ballerina/helpers.bal

```diff
@@ -168,52 +175,83 @@ function addToTable(stream<Type, CompletionType> strm, table<map<Type>> tbl, err
         table<map<Type>> tbl2 = createTableWithKeySpecifier(tbl, typeof(tempTbl));
-        table<map<Type>> tempTable = check createTable(strm, tbl2, err);
+        table<map<Type>> tempTable = check createTable(strm, tbl2);
         return createImmutableTable(tbl, tempTable.toArray());
     }
-    return createTable(strm, tbl, err);
+    return createTable(strm, tbl);
 }
 
-function createTable(stream<Type, CompletionType> strm, table<map<Type>> tbl, error? err) returns table<map<Type>>|error {
+function createTable(stream<Type, CompletionType> strm, table<map<Type>> tbl) returns table<map<Type>>|error {
     record {| Type value; |}|CompletionType v = strm.next();
     while (v is record {| Type value; |}) {
         error? e = trap tbl.add(<map<Type>> checkpanic v.value);
         if (e is error) {
-            if (err is error) {
-                return err;
-            }
             tbl.put(<map<Type>> checkpanic v.value);
         }
         v = strm.next();
     }
-    if (v is error) {
+    if v is error {
         return v;
     }
     return tbl;
 }
 
-function addToMap(stream<Type, CompletionType> strm, map<Type> mp, error? err, boolean isReadOnly) returns map<Type>|error {
-// Here, `err` is used to get the expression of on-conflict clause
+function addToTableForOnConflict(stream<Type, CompletionType> strm, table<map<Type>> tbl, boolean isReadOnly) 
+    returns table<map<Type>>|error {
+    if isReadOnly {
+        // TODO: Properly fix readonly scenario - Issue lang/#36721
+        // In this case tbl will be an immutable table. Therefore, we will create a new mutable table. Next, we will
+        // pass the newly created table into createTableWithKeySpecifier() to add the key specifier details from the
+        // original table variable (tbl). Then the newly created table variable will be populated using createTable()
+        // and make it immutable with createImmutableTable().
+        table<map<Type>> tempTbl = table [];
+        table<map<Type>> mutableTableRef = createTableWithKeySpecifier(tbl, typeof(tempTbl));
+        _ = check createTableForOnConflict(strm, mutableTableRef);
+        return createImmutableTable(tbl, mutableTableRef.toArray());
+    }
+    return createTableForOnConflict(strm, tbl);
+}
+
+function createTableForOnConflict(stream<Type, CompletionType> strm, table<map<Type>> tbl) 
+    returns table<map<Type>>|error {
+    record {| Type value; |}|CompletionType v = strm.next();
+    while v is record {| Type value; |} {
+        record {|Type v; error? err;|}|error value = v.value.ensureType();
+        if value is error {
+            return value;
+        }
+        map<Type> tblValue = check value.v.ensureType();
+        error? e = trap tbl.add(tblValue);
+        error? err = value.err;
+        if e is error && err is error {
+            return err;
+        }
+        if e is error && err is () {
+            tbl.put(<map<Type>> checkpanic value.v);
+        }
```

Comment:
I think it will be more readable if we can use `hasKey`.

Reply:
We cannot use `hasKey` because we don't know the key.